### PR TITLE
update FSDPTransform to delegate materialization to the MaterializationTransform

### PR DESCRIPTION
**Pull request by t-vi**
This is the fourth / final step for #987 .

After this, the FSDP transform plays well with Materialization Transform and should also work with previous transforms (however, the bnb quantization changes tensor layout in a way not terribly amenable to sharding, this needs investigation).

This is a draft to remember to update the branch.

cc: @crcrpar 

*Edit* I need to do my rebasing homework. This PR only changes files in distributed/ the others are from a merge with #980 which will go away after rebasing.

### CODE REVIEW

File: thunder/core/module.py [hunk 1]
```diff
@@ -105,6 +106,18 @@ def named_buffers(self, prefix="", recurse=True, remove_duplicate=True, *, persi
             remove_duplicate=remove_duplicate,
         )
 
+    def _get_shared_names(self):
+        parameters_to_names = WeakTensorKeyDictionary()
+        for name, v in itertools.chain(
+            self.named_parameters(remove_duplicate=False), self.named_buffers(remove_duplicate=False)
+        ):
+            parameters_to_names.setdefault(v, set()).add(name)
+        shared_names = {}
```
Comment (crcrpar) [inline thunder/core/module.py:115]:
```suggestion
        shared_names: dict[str, str] = {}
```

File: thunder/distributed/transforms/fsdp_v2.py [hunk 2]
```diff
@@ -153,92 +153,91 @@ def transform_module(
...
+        # for un-materialized parameters and buffers, we set the ._thunder_device
+
+        is_fully_materialized = True
+        for n, p in thunder_model.named_parameters():
+            for nn in shared_names[n]:
+                if nn != n:
+                    self.shared_params_name[nn] = n
+            try:
+                orig_p = thunder_model._model.get_parameter(n)
+            except AttributeError as _:
```
Comment (crcrpar) [inline thunder/distributed/transforms/fsdp_v2.py:171]:
```suggestion
            except AttributeError:
```

File: thunder/distributed/transforms/fsdp_v2.py [hunk 3]
```diff
@@ -153,92 +153,91 @@ def transform_module(
...
+            elif p.device != self.device:
+                new_p = torch.nn.Parameter(p.to(device=self.device), requires_grad=p.requires_grad)
+                for nn in shared_names[n]:
+                    thunder_model._overrides_parameters[nn] = new_p
+                    device_adjustments[nn] = self.device
+
+        for n, b in thunder_model.named_buffers():
+            try:
+                orig_b = thunder_model._model.get_buffer(n)
+            except AttributeError as _:
```
Comment (crcrpar) [inline thunder/distributed/transforms/fsdp_v2.py:190]:
```suggestion
            except AttributeError:
```

File: thunder/distributed/transforms/fsdp_v2.py [hunk 4]
```diff
@@ -153,92 +153,91 @@ def transform_module(
...
+            if p.is_meta:
+                is_fully_materialized = False
+                p._thunder_device = self.device
+                if orig_p is not None:
+                    orig_p._thunder_device = self.device
+                # TODO: check if devic_adjustments are still needed
+                for nn in shared_names[n]:
+                    device_adjustments[nn] = self.device
+            elif p.device != self.device:
+                new_p = torch.nn.Parameter(p.to(device=self.device), requires_grad=p.requires_grad)
```
Comment (crcrpar) [inline thunder/distributed/transforms/fsdp_v2.py:182]: wouldn't this line want to be in the context of `torch.no_grad()`?

```suggestion
                new_p = torch.nn.Parameter(p.to(device=self.device), requires_grad=p.requires_grad)
```

File: thunder/distributed/transforms/fsdp_v2.py [hunk 5]
```diff
@@ -153,92 +153,91 @@ def transform_module(
...
+                # Note: we could move broadcasting into its own transform coming
+                #       after materialization (in thunder.distributed.fsdp) to
+                #       support this, if it is useful.
+                raise RuntimeError("cannot broadcast from non-materialized model")
+            for pn, p in chain(thunder_model.named_parameters(), thunder_model.named_buffers()):
+                tdist.broadcast(
+                    p,
+                    src=self.broadcast_from,
+                    group=self.process_group,
+                    async_op=False,
```
Comment (crcrpar) [inline thunder/distributed/transforms/fsdp_v2.py:217]: ditto, shouldn't this execution be under `torch.no_grad()`?

File: thunder/distributed/transforms/fsdp_v2.py [hunk 6]
```diff
@@ -153,92 +153,91 @@ def transform_module(
...
+            new_shape = p_new.shape
+            for nn in shared_names[pn]:
+                thunder_model._overrides_parameters[nn] = p_new
+                self.sharded_params[nn] = (old_shape, new_shape, getattr(p, "_thunder_device", p.device))
+            if self.release_original_parameters:
+                p_orig = thunder_model._model.get_parameter(pn)
+                if p_orig.device.type != "meta":
+                    p_meta = torch.nn.Parameter(p_orig.to(device="meta"), requires_grad=p_orig.requires_grad)
+                    p_meta._thunder_device = p_orig.device
+                    for nn in shared_names[pn]:
```
Comment (crcrpar) [inline thunder/distributed/transforms/fsdp_v2.py:235]: what does "nn" stand for? IMHO nn is a common alias for `torch.nn` so I'd love to avoid using this abbreviation

File: thunder/distributed/transforms/fsdp_v2.py [hunk 7]
```diff
@@ -153,92 +153,91 @@ def transform_module(
...
-                    "param_copy": thunder_model._overrides_parameters[pn],
-                    "param_shard_meta": self.sharded_params[pn],
-                    "param_name": pn,
-                }
+
+        # do the actual sharding. Note that meta tensors will give sharded meta tensors.
+        for pn, p in list(thunder_model.named_parameters()):
+            # we collect shapes and devices because we do not know if other transforms also change it.
+            old_shape = p.shape
+            p_new, _ = _shard_tensor(p, global_rank, world_size, pn, allow_padding_for_fsdp=True)
```
Comment (crcrpar) [inline thunder/distributed/transforms/fsdp_v2.py:224]: no_grad?

File: thunder/transforms/materialization.py [hunk 8]
```diff
@@ -55,3 +83,69 @@ def module_init_from_transformed_state_dict(transform: MaterializationTransform,
             model.load_state_dict(state_dict)
 
         return module_init_from_transformed_state_dict
+
+    @staticmethod
+    def init_from_original_module_init():
```
Comment (crcrpar) [inline thunder/transforms/materialization.py:88]: question kind of orthogonal to this pr: would it be possible that one passes sharded state dict to this one?

File: thunder/tests/test_transforms.py [hunk 9]
```diff
@@ -221,3 +221,56 @@ def test_nvfuser_cse():
...
+    from thunder.transforms.quantization import BitsAndBytesLinearQuant4bit, get_bitsandbytes_executor
+
+    bitsandbytes_executor = get_bitsandbytes_executor()
+
+    def get_model():
+        m0 = torch.nn.Linear(2, 2)
+
+        # to not change the rng state
+        with torch.device("meta"):
+            m4 = torch.nn.Linear(2, 2)
```
Comment (crcrpar) [inline thunder/tests/test_transforms.py:238]: with repeated `torch.manual_seed` below, I'm not quite following whether this comment / this use of meta device matter...

File: thunder/transforms/materialization.py (hunk 8 above)
Comment (t-vi) [reply thunder/transforms/materialization.py:88]: If you have a sharded state dict, you should be able to use `module_init_from_transformed_state_dict` (provided you don't have other state-dict-changing transformations afterwards.

File: thunder/tests/test_transforms.py (hunk 9 above)
Comment (t-vi) [reply thunder/tests/test_transforms.py:238]: This is the RNG state within one module initialization (i.e. if this called randn for device tensors, it would put the parameters here that would usually be in the middle layer, the `manual_seed` only sets the seed).
Note that this is part of #980 and just shows here due to a missing rebase.

File: thunder/core/module.py (hunk 1 above)
Comment (t-vi) [reply thunder/core/module.py:115]:
```suggestion
        shared_names: dict[str, set[str]] = {}
```

File: thunder/distributed/transforms/fsdp_v2.py (hunk 4 above)
Comment (t-vi) [reply thunder/distributed/transforms/fsdp_v2.py:182]:
```suggestion
                with torch.no_grad():
                    new_p = torch.nn.Parameter(p.to(device=self.device), requires_grad=p.requires_grad)
```

File: thunder/distributed/transforms/fsdp_v2.py (hunk 7 above)
Comment (t-vi) [reply thunder/distributed/transforms/fsdp_v2.py:224]: Is now on the shard tensor.

File: thunder/distributed/transforms/fsdp_v2.py (hunk 5 above)
Comment (t-vi) [reply thunder/distributed/transforms/fsdp_v2.py:217]: Don't know. Was it before?

File: thunder/distributed/transforms/fsdp_v2.py (hunk 6 above)
Comment (t-vi) [reply thunder/distributed/transforms/fsdp_v2.py:235]: Good call. I'll change it after the rebase.

**Comment by t-vi** [comment]
I need to investigate the failing test in test_transforms in more detail, the FSDP part seems in good shape.

File: thunder/distributed/transforms/fsdp_v2.py (hunk 6 above)
Comment (t-vi) [reply thunder/distributed/transforms/fsdp_v2.py]: fixed this.

File: thunder/distributed/transforms/fsdp_v2.py [hunk 10]
```diff
@@ -156,92 +156,92 @@ def transform_module(
...
+            try:
+                orig_p = thunder_model._model.get_parameter(n)
+            except AttributeError:
+                orig_p = None
+            if p.is_meta:
+                is_fully_materialized = False
+                p._thunder_device = self.device
+                if orig_p is not None:
+                    orig_p._thunder_device = self.device
+                # TODO: check if devic_adjustments are still needed
```
Comment (lantiga) [inline thunder/distributed/transforms/fsdp_v2.py:181]:
```suggestion
                # TODO: check if device_adjustments are still needed
```

**Review by lantiga — APPROVED** [review]
Looks great, although to @crcrpar's point I'm not sure about `no_grad` for `broadcast` - we weren't doing it earlier.

File: thunder/distributed/transforms/fsdp_v2.py (hunk 5 above)
Comment (t-vi) [reply thunder/distributed/transforms/fsdp_v2.py]: added it, thank you!

File: thunder/core/transforms.py [hunk 11]
```diff
@@ -403,7 +403,7 @@ def visitor_transform(trace_from: Trace, visit: Callable, *, provenance: None |
 def add_transform(
     cfn: Callable,
     *,
-    transform: Transform,
+    transform: Transform | list[Transform],
```
[the remark this answers is not in the document]
Comment (lantiga) [reply thunder/core/transforms.py:406]: let's not do that then, totally fine
